src/common/host_device_vector.cu

```diff
@@ -24,8 +24,9 @@ struct HostDeviceVectorImpl {
     : device_(device), on_d_(device >= 0) {
     if (on_d_) {
       dh::safe_cuda(cudaSetDevice(device_));
-      data_d_.resize(init.size());
-      thrust::copy(init.begin(), init.end(), data_d_.begin());
+      if (data_d_ == nullptr) data_d_.reset(new thrust::device_vector<T>());
```

Comment:
`data_d_` is not initialized here, so no need to check it before calling `reset()`.